### ISSUE TITLE
Go To Definition functionality

### ISSUE THREAD

**Issue by Danp2** [genius257/vscode-autoit#24]
Running some basic tests, I've encountered one small issue with the GTD functionality. Here's a code sample --

```autoit
#include "..\webdriver\wd_core.au3"
#include "..\webdriver\wd_capabilities.au3"

Func Main()
	$_WD_Debug = $_WD_DEBUG_Full

	_WD_CapabilitiesDefine($_WD_KEYS__STANDARD_PRIMITIVE, 'webSocketUrl')
	_WD_CapabilitiesDefine($_WD_KEYS__STANDARD_PRIMITIVE, 'moz:debuggerAddress')
EndFunc

```

With the above, everything is good except that I am unable to find a reference for `$_WD_Debug`, which is a global variable declared in wd_core.au3. Taking a deeper dive, it seems that most of the global variables from include files are not being recognized.

**Comment by genius257** [comment]
So this one is a bug, but it is tricky.

When you want to find the declaration for `$_WD_Debug`, the line itself within the function `main` **IS** a variable declaration.
It is however not the **FIRST**...
If you used goto declaration on `$_WD_Debug` used in a expression in main, but after your declaration, would you expect the declaration in main or the first found declaration to be found, and who of the two options is most helpful in general development context?

There is also the question of where global defined variables within functions or conditionals fall into this declaration hierarchy. 

I would argue my solution makes sense in a development environment, but when testing behavior in PHP (A language where no keyword is needed to declare a variable) with the `DEVSENSE.phptools-vscode` extension, it jumps to the first declaration occurrence outside of an function, so this seems to be the solution to implement? :)

**Comment by Sven-Seyfert** [comment]
Hi @genius257,

Sven (SOLVE-SMART) here. In my opinion

> it jumps to the first declaration occurrence outside of an function, so this seems to be the solution to implement? :)

this would be the way to go and how I would expect it. Like you already mentioned, other extensions do exactly this. Let's see what Dan is thinking of 🤝 .

Best regards
Sven

**Comment by Danp2** [comment]
I believe the default should be to locate the first declaration occurrence outside of a function. Maybe you could offer a way to toggle between these two "styles" of GTD.

**Comment by Danp2** [comment]
What should happen in a scenario like the following?

```autoit
Global $x

$x = 'test1'
myFunc1('test2')
MyFunc2()

Func MyFunc1($x)
	ConsoleWrite($x & @CRLF)
EndFunc

Func MyFunc2()
	Local $x

	$x = 'test3'
EndFunc
```

Both functions create a local instance of `$x`, but GTD always refers to the initial Global.

**Comment by Sven-Seyfert** [comment]
I personally would expect, the search for the `Global` occurrence and if there is non then search for a `Local` one.
But nevermind, it's on @genius257 to decide 🤭 .

**Comment by genius257** [comment]
So I'm thinking using the actual AutoIt3 runtime behavior for varaibles in scopes (functions) and how most if not all other language intellisense does it for global.

The following list behavior will stop and return if match is found.

Within a function:

1. check function parameters for a match
2. check lines before the current origin of the declaration request, within the function
3. default to the Global lookup behavior

Outside of a function:

1. Check lines before the current origin of the declaration request.
    1.1. Check for declaration statement
    1.2. Check for include statement and include recursively on the included file, if resolved to an existing file when checking.
2. check for declaration statements with global scope specified within functions.

In step 1 of `Outside of a function` case, function declarations will be added to an array, making it correctly sorted for step 2

Let me know if it makes sense :)

---

I just realized that i currently aren't processing [Assign](https://www.autoitscript.com/autoit3/docs/functions/Assign.htm), [Eval](https://www.autoitscript.com/autoit3/docs/functions/Eval.htm), [Execute](https://www.autoitscript.com/autoit3/docs/functions/Execute.htm) and [IsDeclared](https://www.autoitscript.com/autoit3/docs/functions/IsDeclared.htm).

I plan on making issue for each. They will be low priority for now, just a heads up...

**Comment by genius257** [comment]
Should now be fixed with version 1.2.4 :smile:

**Comment by Danp2** [comment]
@genius257 Unfortunately, the latest update appears to have broken the GTD functionality for functions. ☹️

**Comment by genius257** [comment]
I'm sorry to hear :disappointed: 
I will look into it and fix it! :smile: Hopefully without breaking anything else :sweat_smile: 
Jest tests are not implemented yet, so my competence as tester before release is currently the only safeguard, besides TypeScript type checking :stuck_out_tongue_winking_eye:

**Comment by Danp2** [comment]
Fixed by 254eb03b1603b8d88d3d7253aed8fc23d94c4d52
